### PR TITLE
Add support for cur_pos tensor arg in sdpa decode

### PR DESCRIPTION
**Pull request by sraizada-tt**
### Ticket
#11457

### Problem description
Support cur_pos as a tensor argument to sdpa_decode to enable tracing

### What's changed
Added cur_pos_tensor as an optional argument to ttnn.transformer.scaled_dot_product_attention_decode

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by arakhmati — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/compute/sdpa_flash_decode.cpp [hunk 1]
```diff
@@ -477,15 +479,38 @@ void MAIN {
...
-    const bool do_reduce = get_arg_val<uint32_t>(3) == 1;
+    const bool do_reduce = get_arg_val<uint32_t>(0) == 1;
+    const uint32_t core_num = get_arg_val<uint32_t>(1);
+    const uint32_t cur_batch = get_arg_val<uint32_t>(2);
+    const uint32_t cur_pos_arg = get_arg_val<uint32_t>(3);
+
     // const uin32_t idle_core = get_arg_val<uint32_t>(4);
 
+    // idle core
+    if (get_arg_val<uint32_t>(0)==65){
```
Comment (caixunshiren) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/compute/sdpa_flash_decode.cpp:490]: why is it 65 -- what's the logic here?

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/compute/sdpa_flash_decode.cpp [hunk 2]
```diff
@@ -477,15 +479,38 @@ void MAIN {
...
+        return;
+    }
+
+    // Get cur_pos
+    uint32_t cur_pos = 0;
+    if (cur_pos_arg!=4294967295){
```
Comment (caixunshiren) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/compute/sdpa_flash_decode.cpp:496]: would be good to comment on why the hard-coded values are chosen. Maybe define them as constants?

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/sdpa_decode.cpp [hunk 3]
```diff
@@ -28,14 +28,15 @@ ttnn::Tensor ExecuteScaledDotProductAttentionDecode::invoke(
...
     std::optional<float> scale,
     const std::optional<MemoryConfig> &memory_config,
     std::optional<SDPAProgramConfig> program_config,
     std::optional<DeviceComputeKernelConfig> compute_kernel_config) {
     auto arch = input_tensor_q.storage_type() == StorageType::DEVICE ? input_tensor_q.device()->arch()
                                                                      : AutoFormat::GetDefaultDevice()->arch();
-    uint32_t max_cur_pos = *std::max_element(cur_pos.begin(), cur_pos.end());
-    uint32_t k_chunk_size = get_chunk_size(max_cur_pos + 1);
+    //uint32_t max_cur_pos = *std::max_element(cur_pos.begin(), cur_pos.end());
+    uint32_t k_chunk_size = 512; //get_chunk_size(max_cur_pos + 1);
```
Comment (caixunshiren) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_decode/sdpa_decode.cpp:39]: do we always use chunk size 512 now? what's the reason?

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/dataflow/reader_decode_all.cpp [hunk 4]
```diff
@@ -26,19 +26,50 @@ void kernel_main() {
...
+    const uint32_t core_num = get_arg_val<uint32_t>(6);
+    const uint32_t cur_pos_arg = get_arg_val<uint32_t>(7);
+
+    // idle core
+    if (q_addr ==0){
+        return;
+    }
+    // Get cur_pos
+    uint32_t cur_pos = 0;
+    if (cur_pos_arg!=4294967295){
```
Comment (caixunshiren) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/dataflow/reader_decode_all.cpp:49]: same comment as above

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/dataflow/writer_decode_all.cpp [hunk 5]
```diff
@@ -261,19 +262,36 @@ void kernel_main() {
...
+    const uint32_t core_num = get_arg_val<uint32_t>(4);
+    const uint32_t cur_pos_arg = get_arg_val<uint32_t>(5);
+
+    // idle core
+    if (out_addr == 0){
+        return;
+    }
+    // Get cur_pos
+    uint32_t cur_pos = 0;
+    if (cur_pos_arg!=4294967295){
```
Comment (caixunshiren) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/dataflow/writer_decode_all.cpp:280]: ^^

**Review by caixunshiren — CHANGES_REQUESTED** [review]
Overall, amazing work!

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/sdpa_decode.cpp (hunk 3 above)
Comment (sraizada-tt) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_decode/sdpa_decode.cpp:39]: can't do std::max_element(cur_pos.begin(), cur_pos.end()); now

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/compute/sdpa_flash_decode.cpp (hunk 2 above)
Comment (sraizada-tt) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_decode/device/kernels/compute/sdpa_flash_decode.cpp:496]: end of uint32 range

**Review by caixunshiren — APPROVED** [review]

**Pull request merged**